src/engine.ml

```diff
@@ -183,6 +183,61 @@ let client config ts now rng =
         let remotes = Array.to_list remotes in
         Config.add Remote remotes config
   in
+  let* config =
+    match Config.find Pkcs12 config with
+    | None -> Ok config
+    | Some p12 -> (
+        let* pass =
+          Option.to_result ~none:(`Msg "missing pkcs12 password")
+            pkcs12_password
+        in
+        let* stuff = X509.PKCS12.verify pass p12 in
+        let certs =
+          List.filter_map
+            (function `Certificate c -> Some c | _ -> None)
+            stuff
+        and keys =
+          List.filter
+            (function
+              | `Private_key _ | `Decrypted_private_key _ -> true | _ -> false)
+            stuff
+        in
+        match keys with
+        | [ (`Private_key pk | `Decrypted_private_key pk) ] -> (
+            let key_fp =
+              X509.Public_key.fingerprint (X509.Private_key.public pk)
+            in
+            match
+              List.partition
+                (fun cert ->
+                  let cert_pubkey = X509.Certificate.public_key cert in
+                  Cstruct.equal (X509.Public_key.fingerprint cert_pubkey) key_fp)
+                certs
+            with
+            | [ cert ], extra_certs ->
+                (* OpenVPN adds the certs to [Ca] if it is not present;
+                   we don't as it is easily surprising behavior *)
+                if extra_certs <> [] then
+                  Log.warn (fun m ->
+                      m "Extra certificates found in PKCS12 file%s: %a"
+                        (if Config.mem Ca config then ""
```

Comment:
I'm not sure our config validator allows for `--pkcs12` without `--ca` hmm.

Reply:
it does, ca is completely optional:
```OCaml
          match Config.find Ca config with
          | None ->
              Log.warn (fun m ->
                  m "not authenticating certificate (missing CA)");
              fun ?ip:_ ~host:_ _ -> Ok None
          | Some ca ->
              Log.info (fun m ->
                  m "authenticating with CA %a"
                    Fmt.(list ~sep:(any "\n") X509.Certificate.pp)
                    ca);
```